File: MacOS/syncFolder/App/AppState.swift
import Foundation
import SwiftUI

@MainActor
final class AppState: ObservableObject {
    // MARK: - Published State

    @Published var pairs: [SyncPair] = []
    @Published var statuses: [UUID: PairStatus] = [:]
    @Published var logs: [LogEntry] = []
    @Published var notificationsEnabled: Bool {
        didSet { UserDefaults.standard.set(notificationsEnabled, forKey: "notificationsEnabled") }
    }

    // MARK: - Computed

    var isSyncing: Bool {
        statuses.values.contains { $0.state == .syncing }
    }

    var activePairCount: Int {
        pairs.filter(\.isEnabled).count
    }

    // MARK: - Private

    private let configManager = ConfigManager.shared
    private let engine = SyncEngine()
    private var scheduledTasks: [UUID: Task<Void, Never>] = [:]

    // MARK: - Init

    init() {
        self.notificationsEnabled = UserDefaults.standard.bool(forKey: "notificationsEnabled")
        pairs = configManager.loadPairs()
        if notificationsEnabled {
            NotificationManager.shared.requestPermission()
        }
        scheduleAll()
    }

    // MARK: - Pair Management

    func addPair(_ pair: SyncPair) {
        pairs.append(pair)
        save()
    }

    func updatePair(_ pair: SyncPair) {
        if let i = pairs.firstIndex(where: { $0.id == pair.id }) {
            pairs[i] = pair
            save()
        }
    }

    func deletePair(_ pair: SyncPair) {
        // Cancel the task first, then clean up
        scheduledTasks[pair.id]?.cancel()
        scheduledTasks.removeValue(forKey: pair.id)
        pairs.removeAll { $0.id == pair.id }
        statuses.removeValue(forKey: pair.id)
        ManifestManager.shared.deleteManifest(pairId: pair.id)
        configManager.savePairs(pairs)
    }

    func save() {
        configManager.savePairs(pairs)
        scheduleAll()
    }

    // MARK: - Sync

    func syncNow(pairId: UUID? = nil) {
        let pairsToSync: [SyncPair]
        if let id = pairId {
            pairsToSync = pairs.filter { $0.id == id }
        } else {
            pairsToSync = pairs.filter(\.isEnabled)
        }
        for pair in pairsToSync {
            Task { await performSync(pair: pair) }
        }
    }

    private func performSync(pair: SyncPair) async {
        // Skip if already syncing this pair
        guard statuses[pair.id]?.state != .syncing else { return }

        // Mark as syncing (preserving last result for UI)
        let prev = statuses[pair.id]
        statuses[pair.id] = PairStatus(state: .syncing,
                                       lastResult: prev?.lastResult,
                                       lastSyncDate: prev?.lastSyncDate)

        let result: SyncResult
        do {
            result = try await engine.sync(pair: pair)
            statuses[pair.id] = PairStatus(state: .idle, lastResult: result, lastSyncDate: Date())
        } catch {
            result = SyncResult(filesCopied: 0, filesDeleted: 0, bytesTransferred: 0,
                                errors: [error.localizedDescription])
            statuses[pair.id] = PairStatus(state: .error,
                                           errorMessage: error.localizedDescription,
                                           lastResult: result,
                                           lastSyncDate: Date())
        }

        appendLog(LogEntry(pairId: pair.id, pairName: pair.name, result: result, date: Date()))

        // Send notifications
        if notificationsEnabled {
            if statuses[pair.id]?.state == .error {
                NotificationManager.shared.sendSyncFailed(
                    pairName: pair.name,
                    error: result.errors.first ?? "Unknown error"
                )
            } else if !result.isClean {
                NotificationManager.shared.sendSyncCompleted(pairName: pair.name, result: result)
            }
        }

        // Persist last sync date
        if let i = pairs.firstIndex(where: { $0.id == pair.id }) {
            pairs[i].lastSyncDate = Date()
            configManager.savePairs(pairs)
        }
    }

    // MARK: - Scheduler

    func scheduleAll() {
        // Cancel all existing scheduled tasks
        for (_, task) in scheduledTasks {
            task.cancel()
        }
        scheduledTasks.removeAll()

        for pair in pairs where pair.isEnabled {
            schedulePair(pair)
        }
    }

    private func schedulePair(_ pair: SyncPair) {
        let intervalSeconds = pair.intervalMinutes * 60
        let task = Task { [weak self] in
            // Small delay before first sync so UI has time to settle
            try? await Task.sleep(for: .seconds(3))
            guard !Task.isCancelled else { return }

            await self?.performSync(pair: pair)

            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(intervalSeconds))
                guard !Task.isCancelled else { break }
                await self?.performSync(pair: pair)
            }
        }
        scheduledTasks[pair.id] = task
    }

    // MARK: - Logs

    private func appendLog(_ entry: LogEntry) {
        logs.insert(entry, at: 0)
        if logs.count > 200 {
            logs = Array(logs.prefix(200))
        }
    }

    func clearLogs() {
        logs.removeAll()
    }
}

File: Windows/syncFolder/ViewModels/AppState.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using syncFolder.Models;
using syncFolder.Services;

namespace syncFolder.ViewModels;

public class AppState : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;

    private readonly ConfigManager _configManager = ConfigManager.Shared;
    private readonly SyncEngine _engine = new();
    private readonly Dictionary<Guid, CancellationTokenSource> _scheduledTasks = new();

    // Observable collections for UI binding
    public ObservableCollection<SyncPair> Pairs { get; } = new();
    public ObservableCollection<LogEntry> Logs { get; } = new();
    public Dictionary<Guid, PairStatus> Statuses { get; } = new();

    private bool _notificationsEnabled;
    public bool NotificationsEnabled
    {
        get => _notificationsEnabled;
        set
        {
            if (_notificationsEnabled != value)
            {
                _notificationsEnabled = value;
                Properties.Settings.Default.NotificationsEnabled = value;
                Properties.Settings.Default.Save();
                OnPropertyChanged();
            }
        }
    }

    public bool IsSyncing => Statuses.Values.Any(s => s.State == PairStatus.StateType.Syncing);
    public int ActivePairCount => Pairs.Count(p => p.IsEnabled);
    public int ErrorCount => Statuses.Values.Count(s => s.State == PairStatus.StateType.Error);

    public AppState()
    {
        _notificationsEnabled = Properties.Settings.Default.NotificationsEnabled;
        var loaded = _configManager.LoadPairs();
        foreach (var pair in loaded)
            Pairs.Add(pair);

        ScheduleAll();
    }

    // MARK: - Pair Management

    public void AddPair(SyncPair pair)
    {
        Pairs.Add(pair);
        Save();
    }

    public void UpdatePair(SyncPair pair)
    {
        for (int i = 0; i < Pairs.Count; i++)
        {
            if (Pairs[i].Id == pair.Id)
            {
                Pairs[i] = pair;
                break;
            }
        }
        Save();
    }

    public void DeletePair(SyncPair pair)
    {
        if (_scheduledTasks.TryGetValue(pair.Id, out var cts))
        {
            cts.Cancel();
            _scheduledTasks.Remove(pair.Id);
        }

        Pairs.Remove(pair);
        Statuses.Remove(pair.Id);
        ManifestManager.Shared.DeleteManifest(pair.Id);
        _configManager.SavePairs(new List<SyncPair>(Pairs));
        NotifyStatusChanged();
    }

    public void Save()
    {
        _configManager.SavePairs(new List<SyncPair>(Pairs));
        ScheduleAll();
    }

    // MARK: - Sync

    public void SyncNow(Guid? pairId = null)
    {
        var pairsToSync = pairId.HasValue
            ? Pairs.Where(p => p.Id == pairId.Value).ToList()
            : Pairs.Where(p => p.IsEnabled).ToList();

        foreach (var pair in pairsToSync)
        {
            _ = PerformSyncAsync(pair);
        }
    }

    private async Task PerformSyncAsync(SyncPair pair)
    {
        // Skip if already syncing
        if (Statuses.TryGetValue(pair.Id, out var currentStatus) && currentStatus.State == PairStatus.StateType.Syncing)
            return;

        // Mark as syncing
        var prev = Statuses.GetValueOrDefault(pair.Id);
        Statuses[pair.Id] = new PairStatus
        {
            State = PairStatus.StateType.Syncing,
            LastResult = prev?.LastResult,
            LastSyncDate = prev?.LastSyncDate
        };
        NotifyStatusChanged();

        SyncResult result;
        try
        {
            result = await _engine.SyncAsync(pair);
            Statuses[pair.Id] = new PairStatus
            {
                State = PairStatus.StateType.Idle,
                LastResult = result,
                LastSyncDate = DateTime.Now
            };
        }
        catch (Exception ex)
        {
            result = new SyncResult
            {
                FilesCopied = 0,
                FilesDeleted = 0,
                BytesTransferred = 0,
                Errors = new List<string> { ex.Message }
            };
            Statuses[pair.Id] = new PairStatus
            {
                State = PairStatus.StateType.Error,
                ErrorMessage = ex.Message,
                LastResult = result,
                LastSyncDate = DateTime.Now
            };
        }

        AppendLog(new LogEntry(pair.Id, pair.Name, result, DateTime.Now));
        NotifyStatusChanged();

        // Send notifications
        if (NotificationsEnabled)
        {
            if (Statuses[pair.Id].State == PairStatus.StateType.Error)
            {
                NotificationManager.Shared.SendSyncFailed(pair.Name, result.Errors.FirstOrDefault() ?? "Unknown error");
            }
            else if (!result.IsClean)
            {
                NotificationManager.Shared.SendSyncCompleted(pair.Name, result);
            }
        }

        // Persist last sync date
        var found = Pairs.FirstOrDefault(p => p.Id == pair.Id);
        if (found != null)
        {
            found.LastSyncDate = DateTime.Now;
            _configManager.SavePairs(new List<SyncPair>(Pairs));
        }
    }

    // MARK: - Scheduler

    public void ScheduleAll()
    {
        foreach (var cts in _scheduledTasks.Values)
            cts.Cancel();
        _scheduledTasks.Clear();

        foreach (var pair in Pairs.Where(p => p.IsEnabled))
            SchedulePair(pair);
    }

    private void SchedulePair(SyncPair pair)
    {
        var cts = new CancellationTokenSource();
        _scheduledTasks[pair.Id] = cts;
        var token = cts.Token;
        var intervalMs = pair.IntervalMinutes * 60 * 1000;

        _ = Task.Run(async () =>
        {
            try
            {
                // Small delay before first sync
                await Task.Delay(3000, token);
                if (token.IsCancellationRequested) return;

                // Dispatch to UI thread for sync
                await System.Windows.Application.Current.Dispatcher.InvokeAsync(() => PerformSyncAsync(pair));

                while (!token.IsCancellationRequested)
                {
                    await Task.Delay(intervalMs, token);
                    if (token.IsCancellationRequested) break;
                    await System.Windows.Application.Current.Dispatcher.InvokeAsync(() => PerformSyncAsync(pair));
                }
            }
            catch (TaskCanceledException)
            {
                // Expected when cancelling
            }
        }, token);
    }

    // MARK: - Logs

    private void AppendLog(LogEntry entry)
    {
        System.Windows.Application.Current.Dispatcher.Invoke(() =>
        {
            Logs.Insert(0, entry);
            while (Logs.Count > 200)
                Logs.RemoveAt(Logs.Count - 1);
        });
    }

    public void ClearLogs()
    {
        Logs.Clear();
        OnPropertyChanged(nameof(Logs));
    }

    // MARK: - Notifications

    public void NotifyStatusChanged()
    {
        OnPropertyChanged(nameof(IsSyncing));
        OnPropertyChanged(nameof(ActivePairCount));
        OnPropertyChanged(nameof(ErrorCount));
        OnPropertyChanged(nameof(Statuses));
    }

    protected void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    public void Shutdown()
    {
        foreach (var cts in _scheduledTasks.Values)
            cts.Cancel();
        _scheduledTasks.Clear();
    }
}
